[dtensor] add foreach_zero_ support

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113897

This PR add foreach_zero_ op support, to fix when
optim.zero_grad(set_to_none=False) hit this op and erroring out the
device mesh not found issue.

Also move the test to use zero_grad as the last step as that's when we
going to have dtensor as grads

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

wanchaol committed:
[dtensor] add foreach_zero_ support

This PR add foreach_zero_ op support, to fix when
optim.zero_grad(set_to_none=False) hit this op and erroring out the
device mesh not found issue.

Also move the test to use zero_grad as the last step as that's when we
going to have dtensor as grads

[ghstack-poisoned]

awgu commented on test/distributed/_tensor/test_optimizers.py:
```diff
@@ -60,19 +60,19 @@ def _assert_optimizer(
         loss = out.sum()
         loss.backward()
         optim.step()
+        optim.zero_grad(set_to_none=False)
```
I usually do something like `zero_grad(set_to_none=(iter_idx % 2 == 0))` to exercise both paths. It could be worth a try if you want to still test the `zero_grad(True)` case?

awgu replied:
It does not matter much. We can just leave it as is and merge.

wanchaol replied:
sure, good suggestion!

awgu reviewed (approved):
Thank you!

wanchaol committed:
Update on "[dtensor] add foreach_zero_ support"


This PR add foreach_zero_ op support, to fix when
optim.zero_grad(set_to_none=False) hit this op and erroring out the
device mesh not found issue.

Also move the test to use zero_grad as the last step as that's when we
going to have dtensor as grads

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
